maths/harshad_numbers.py

```diff
@@ -0,0 +1,133 @@
+"""
+A harshad number (or more specifically an n-harshad number) is a number that's
+divisible by the sum of its digits in some given base n.
+Reference: https://en.wikipedia.org/wiki/Harshad_number
+"""
+
+
+def int_to_base(number: int, base: int) -> str:
+    digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    result = ""
+
+    if number < 0:
+        raise ValueError("number must be a positive integer")
+
+    while number > 0:
+        number, remainder = divmod(number, base)
+        result = digits[remainder] + result
+
+    if result == "":
+        result = "0"
+
+    return result
+
+
+def sum_of_digits(num: int, base: int) -> str:
+    """
+    Calculate the sum of digit values in a positive integer
+    converted to the given 'base'.
+    Where 'base' ranges from 2 to 36.
+
+    Examples:
+    >>> sum_of_digits(103, 12)
+    '13'
+    >>> sum_of_digits(1275, 4)
+    '30'
+    >>> sum_of_digits(6645, 2)
+    '1001'
+    >>> # bases beyond 36 and below 2 will error
+    >>> sum_of_digits(543, 1)
+    Traceback (most recent call last):
+        ...
+    ValueError: 'base' must be between 36 and 2 inclusive
+    >>> sum_of_digits(543, 37)
+    Traceback (most recent call last):
+        ...
+    ValueError: 'base' must be between 36 and 2 inclusive
+    """
+
+    if (base > 36) or (base < 2):
+        raise ValueError("'base' must be between 36 and 2 inclusive")
+
+    num_str = int_to_base(num, base)
+    res = 0
+    for char in num_str:
+        res += int(char, base)
+    res_str = int_to_base(res, base)
+    return res_str
+
+
+def harshad_numbers_in_base(limit: int, base: int) -> list[str]:
+    """
+    Finds all Harshad numbers smaller than num in base 'base'.
+    Where 'base' ranges from 2 to 36.
+
+    Examples:
+    >>> harshad_numbers_in_base(15, 2)
+    ['1', '10', '100', '110', '1000', '1010', '1100']
+    >>> harshad_numbers_in_base(12, 34)
+    ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B']
+    >>> harshad_numbers_in_base(12, 4)
+    ['1', '2', '3', '10', '12', '20', '21']
+    >>> # bases beyond 36 and below 2 will error
+    >>> harshad_numbers_in_base(234, 37)
+    Traceback (most recent call last):
+        ...
+    ValueError: 'base' must be between 36 and 2 inclusive
+    >>> harshad_numbers_in_base(234, 1)
+    Traceback (most recent call last):
+        ...
+    ValueError: 'base' must be between 36 and 2 inclusive
+    """
+
+    if (base > 36) or (base < 2):
+        raise ValueError("'base' must be between 36 and 2 inclusive")

```

Comment:
```suggestion
    if base < 2 or base > 36:
        raise ValueError("'base' must be between 2 and 36 inclusive")
```
See previous comment